std/async/pool.ts

```diff
@@ -0,0 +1,46 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+
+/**
+ * pooledMap transforms values from an (async) iterable into another async
+ * iterable. The transforms are done concurrently, with a max concurrency
+ * defined by the poolLimit.
+ * 
+ * @param poolLimit The maximum count of items being processed concurrently. 
+ * @param array The input array for mapping.
+ * @param iteratorFn The function to call for every item of the array.
+ */
+export function pooledMap<T, R>(
+  poolLimit: number,
+  array: Iterable<T> | AsyncIterable<T>,
+  iteratorFn: (data: T) => Promise<R>,
+): AsyncIterableIterator<R> {
```

Comment:
The point of using a generator is that there doesn't need to be a mapping function parameter anymore. That would go in the loop body. The function just yields the promised values in order, capping the amount run at once.

Reply:
So a user would write something like this?

```ts
const results = []
for await (const url of pool(2, urls)) {
  results.push(fetch(url))
}
console.log(await Promise.all(results))
```

I guess that would work.

Reply:
Yeah, I just noticed that the loop body is unable to act on different elements at the same time, is that what you meant? There's no meaning in it returning an iterator then... the original definition was fine.

Reply:
Yeah that is what I meant. I think the async iterator still makes sense though - it is the more raw API that lets you stream the responses. By returning a `Promise<R[]>` we are forcing you to use all results at once which is likely not always what you want. You can still get a `Promise<R[]>` from the `AsyncIterableIterator<R>` (really easy once iterator helpers proposal lands).

Reply:
There is this related data structure DeferredStack https://github.com/deno-postgres/deno-postgres/blob/28249fbec10f7f1581f405a5a5b4b105a838629b/deferred.ts

basically you hand over a T when one is available (in deno-postgres T is a Client, essentially a db connection), and you can do something like:
```ts
const results = []
const pool = DeferredStack<undefined>(2, [undefined, undefined])  // here T is undefined
for await (const url of urls)) {
  await pool.pop()
  fetch(url)
    .then(x => results.push(x))
    // "release"
    .finally(x => pool.push(undefined))
}
// note this may be out of order
console.log(await Promise.all(results))
```

This datastructure allows `pool.query(sql)` to hide the connection being locked/released.

It feels like there should be a nice abstraction to this (with a clean `for await` api)...